data-reconciliation-app/docs/adr/02-data-schema-strategy.md

```diff
@@ -0,0 +1,127 @@
+# Data Schema Strategy
+
+## Status
+
+Decided
+
+## Context
+
+We need to build an API that ingests member data into our system. Once ingested, we also need to store this data in the K-V store. Members will decide on a data schema which will be used for ingesting, data reconciliation and reporting.
+
+There are multiple strategies for data schema definition and usage. In this document, we lay out our decision and some other alternatives considered.
+
+## Alternatives
+
+### Option1: Fully Flexible Data Schema
+
+In this strategy, members define their data schema and submit it as part of data ingestion. The API will take the schema and map it to an internal universal schema before storing in the K-V store. When a member requests a data reconciliation report, the API will generate the report using the internal universal schema. This will result in member to understand the schema of the report and possibly convert it back to their custom format to use at their side.
+
+Data ingestion API payload could become:
+
+```json
+POST /submit-data
+
+{
+    "data": [
+        {
+            "key": "abcde",
+            "value": "12345"
+        }
+    ],
+    "schema": {
+        "key": "id",
+        "value": "value"
+    }
+}
+```
+
+This option requires a member to reset their previously submitted data and resubmit again when they change their schema.
+
+While this option provides the maximum flexibility for members to submit data in any form, it introduces data validation and mapping, report generation challenges in every layer at the API.
+
+### Option2: Defining Data Schema via API
```

Comment:
I also wonder if governance could come into play here. Assuming a schema is NOT pre-defined by members, a member could propose a data set(schema) for all members to `vote` on and provide the proposed schema. 

If members agree to that schema, they could then submit opinions/votes on that schema. Just a future thought...

Reply:
yeah, we can also iterate towards that direction. I think it could be more like option 2 implemented using governance mechanism.